File: agent/runtimes/agent-sandbox/clients/python/agentic-sandbox-client/k8s_agent_sandbox/sandbox.py
# Copyright 2026 The Kubernetes Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import atexit
import logging
import requests
from .trace_manager import create_tracer_manager, trace_span, trace
from .commands.command_executor import CommandExecutor
from .files.filesystem import Filesystem
from .models import (
    SandboxConnectionConfig,
    SandboxLocalTunnelConnectionConfig,
    SandboxTracerConfig,
)
from .k8s_helper import K8sHelper
from .connector import SandboxConnector
from .constants import POD_NAME_ANNOTATION, SANDBOX_NAME_HASH_LABEL
from .utils import select_pod_ip

class Sandbox:
    """
    Represents a connection to a specific running Sandbox instance.

    This class provides the interface for interacting with the Sandbox, including:
    - Executing commands via the `commands` property.
    - Managing files via the `files` property.
    - Handling the underlying connection.
    - Integrating with OpenTelemetry for tracing operations.
    """
    def __init__(
        self,
        claim_name: str,
        sandbox_id: str,
        namespace: str = "default",
        connection_config: SandboxConnectionConfig | None = None,
        tracer_config: SandboxTracerConfig | None = None,
        k8s_helper: K8sHelper | None = None,
    ):
        # Sandbox Related Configuration
        self.claim_name = claim_name
        self.sandbox_id = sandbox_id
        self.namespace = namespace
        self.connection_config = connection_config or SandboxLocalTunnelConnectionConfig()

        # Sandbox Management downstream dependency
        self.k8s_helper = k8s_helper or K8sHelper()

        self.connector = SandboxConnector(
            sandbox_id=self.sandbox_id,
            namespace=self.namespace,
            connection_config=self.connection_config,
            k8s_helper=self.k8s_helper,
            get_pod_ip=self.get_pod_ip,
        )

        # Tracer initialization
        self.tracer_config = tracer_config or SandboxTracerConfig()
        self.trace_service_name = self.tracer_config.trace_service_name
        self.tracing_manager, self.tracer = create_tracer_manager(self.tracer_config)

        # Initialisation of namespaced engines
        self._commands = CommandExecutor(self.connector, self.tracer, self.trace_service_name)
        self._files = Filesystem(self.connector, self.tracer, self.trace_service_name)
        
        # Internal state tracking
        self._is_closed = False
        self._pod_name = None
        self._sandbox_name_hash = None
        
    def get_pod_name(self) -> str:
        """Fetches the Sandbox object from Kubernetes and retrieves its current pod name."""
        if self._pod_name is not None:
            return self._pod_name

        sandbox_object = self.k8s_helper.get_sandbox(self.sandbox_id, self.namespace) or {}
        metadata = sandbox_object.get('metadata') or {}
        annotations = metadata.get('annotations') or {}
        pod_name = annotations.get(POD_NAME_ANNOTATION)
        self._pod_name = pod_name if pod_name is not None else self.sandbox_id
        return self._pod_name


    def get_sandbox_name_hash(self) -> str | None:
        """Fetches the Sandbox object from Kubernetes and retrieves its name hash from selector.
        Caches the result to avoid repeated API calls.
        """
        # Return cached value if available
        if self._sandbox_name_hash is not None:
            return self._sandbox_name_hash

        sandbox_object = self.k8s_helper.get_sandbox(self.sandbox_id, self.namespace) or {}
        status = sandbox_object.get('status') or {}
        selector = status.get('selector') or ""
        if "=" in selector:
            key, value = selector.split("=")
            if key == SANDBOX_NAME_HASH_LABEL:
                self._sandbox_name_hash = value
                return value
                
        return None

    def get_pod_ip(self) -> str | None:
        """Selects a pod IP from the Sandbox status (prefers IPv4, normalizes canonical form).

        Always queries the K8s API for the latest IP — the pod IP can change
        after a pod restart (e.g. when spec.operatingMode is set to Suspended and resumed
        via setting spec.operatingMode to Running).
        Returns None if no valid IP can be selected.
        """
        sandbox_object = self.k8s_helper.get_sandbox(self.sandbox_id, self.namespace) or {}
        status_data = sandbox_object.get("status") or {}
        pod_ips = status_data.get('podIPs', [])
        return select_pod_ip(pod_ips)

    def status(self) -> tuple[str, str]:
        """
        Retrieves the current status of the Sandbox by inspecting its Kubernetes conditions.
        
        Returns a tuple of (status, message).
        status can be 'SandboxReady', 'SandboxNotFound', or 'SandboxNotReady'.
        message contains the Kubernetes condition message if available.
        """
        sandbox_object = self.k8s_helper.get_sandbox(self.sandbox_id, self.namespace)
        if not sandbox_object:
            return "SandboxNotFound", "Sandbox object not found in Kubernetes."

        status_data = sandbox_object.get("status") or {}
        for cond in status_data.get("conditions") or []:
            if cond.get("type") == "Ready":
                message = cond.get("message", "")
                if cond.get("status") == "True":
                    return "SandboxReady", message
                else:
                    return "SandboxNotReady", message

        return "SandboxNotReady", "Unknown message"

    @property
    def commands(self) -> CommandExecutor | None:
        return self._commands

    @property
    def files(self) -> Filesystem | None:
        return self._files

    @property
    def is_active(self) -> bool:
        """
        Returns True if the connection hasn't been explicitly closed 
        and engines are still initialized.
        """
        return not self._is_closed and self._commands is not None and self._files is not None

    def close_connection(self):
        """
        Closes the client-side connection and disables execution engines locally,
        but leaves the remote Kubernetes Sandbox infrastructure running.
        
        Use this to free up local resources (like port-forwards or HTTP sessions).
        """
        if self._is_closed:
            return
        # Close client side connection
        self.connector.close()
        
        # Don't allow anymore further executions.
        self._commands = None
        self._files = None
        
        # Cleanup Trace if it exists
        if self.tracing_manager:
            try:
                self.tracing_manager.end_lifecycle_span()
            except Exception as e:
                logging.error(f"Failed to end tracing span: {e}")
        
        self._is_closed = True
        logging.info(f"Connection to sandbox claim '{self.claim_name}' has been closed.")
    
    def terminate(self):
        """
        Permanent deletion of all server side infrastructure and client side connection.

        This method is idempotent. Calling ``terminate()`` repeatedly after a
        successful deletion is a safe no-op. If the remote infrastructure has
        already been removed, subsequent calls will handle the API 404 gracefully
        rather than raising an error.
        """
        # Close the client side connection and trace manager lifecycle
        self.close_connection()

        if not self.claim_name:
            # Already deleted (or never successfully created a claim).
            return

        self.k8s_helper.delete_sandbox_claim(self.claim_name, self.namespace)

        # Clear after successful delete so a retry does not 404.
        self.claim_name = None

File: agent/runtimes/agent-sandbox/test/e2e/framework/predicates/sandbox.go
// Copyright 2025 The Kubernetes Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package predicates

import (
	"fmt"

	"github.com/google/go-cmp/cmp"
	"github.com/google/go-cmp/cmp/cmpopts"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	sandboxv1beta1 "sigs.k8s.io/agent-sandbox/api/v1beta1"
	"sigs.k8s.io/controller-runtime/pkg/client"
)

func asSandbox(obj client.Object) (*sandboxv1beta1.Sandbox, error) {
	if obj == nil {
		return nil, fmt.Errorf("sandbox object is nil")
	}
	sandbox, err := asTyped[*sandboxv1beta1.Sandbox](obj)
	if err != nil {
		return nil, err
	}
	return sandbox, nil
}

// SandboxHasStatus verifies that the Sandbox object has the specified status.
func SandboxHasStatus(status sandboxv1beta1.SandboxStatus) ObjectPredicate {
	return &sandboxHasStatusPredicate{
		WantStatus: status,
	}
}

type sandboxHasStatusPredicate struct {
	WantStatus sandboxv1beta1.SandboxStatus
}

func (s *sandboxHasStatusPredicate) String() string {
	return fmt.Sprintf("SandboxHasStatus(%v)", s.WantStatus)
}

func (s *sandboxHasStatusPredicate) Matches(obj client.Object) (bool, error) {
	sandbox, err := asSandbox(obj)
	if err != nil {
		return false, err
	}
	opts := []cmp.Option{
		cmpopts.IgnoreFields(metav1.Condition{}, "LastTransitionTime"),
		cmpopts.IgnoreFields(sandboxv1beta1.SandboxStatus{}, "PodIPs", "NodeName"),
		// Condition order carries no meaning: conditions are addressed by type, and
		// a condition removed and later re-added is appended at the end, so the
		// order reflects history rather than state. Compare as a set by type.
		cmpopts.SortSlices(func(a, b metav1.Condition) bool { return a.Type < b.Type }),
	}
	if diff := cmp.Diff(s.WantStatus, sandbox.Status, opts...); diff != "" {
		return false, nil
	}
	return true, nil
}
